### DISCUSSION TITLE
NER: replacing the TransitionBasedParser ?

### DISCUSSION THREAD

**Discussion by Stannislav** [explosion/spaCy#7995] [Help: Coding & Implementations] [answered]
Hello community.

# Question
**What is the correct approach of defining a custom NER model that could replace the default `TransitionBasedParser`?**

I have two different approaches in mind that I'm not sure about:

## 1. Re-use the `ner` pipeline component, define custom model
Design a custom NER model and register it via the `@spacy.registry.models("MyCustomNERModel")` decorator. That would correspond to a config file that looks something like that
```ini
# config.cfg
[nlp]
lang = "en"
pipeline = ["tok2vec","ner"]

[components]

[components.ner]
factory = "ner"
moves = null
update_with_oracle_cut_size = 100

[components.ner.model]
@architectures = "MyCustomNERModel"

...
```
## 2. Define a new pipeline component
Design a custom pipeline component and register it similarly to what's done in `spacy/pipeline/ner.pyx` via the decorator `@Language.factory("my_ner", ...)`. The config would probably look like something like this:
```ini
# config.cfg
[nlp]
lang = "en"
pipeline = ["tok2vec","my_ner"]

[components]

[components.my_ner]
factory = "my_ner"
...
```

## Discussion
In `spacy/pipeline/ner.pyx` I see
```python
@Language.factory("ner", ...)
def make_ner(...):
    """Create a transition-based EntityRecognizer component. The entity recognizer
    identifies non-overlapping labelled spans of tokens.
    ...
    """
```
and the docstring above makes me think that the pre-defined "ner" pipeline componenet only works with the transition-based parser. Does that mean approach 1. wouldn't work and I have to go with approach 2.?

# Minimal Working Example?
What I'm trying to come up with is a minimal working example with the following characteristics
- Transformer component produces word embedding vectors
- Custom NER: linear layer + softmax that maps word embeddings to named entity predictions

Is it possible to outline how to approach this?

# Question about the Transition-Based Parser
The current default model for NER is based on the model proposed in [Lample et al.](https://arxiv.org/abs/1603.01360) in 2016.

Given that it's already 5 years old I'm curious why there isn't a more recent model. What is the main motivation?
- This is still state of the art?
- It's fast?
- It can be re-used for both NER and dependency parsing?
- It's easy to implement checks for invalid IOB sequences?
- Something else?

# Thanks a lot!

**Comment by svlandeg** [comment, marked as answer]
Hi!

The `ner` component is powered by an `EntityRecognizer` object which inherits from `Parser` (as does `DependencyParser`). I don't think we've tried changing the `model` instance to something entirely different, and I'm afraid it might not be very straightforward. There are internal references such as using `self.model.attrs["unseen_classes"]`, the loss function is tailored towards the kind of ML model, and all sorts of inter-dependencies are probably implicit. I do think you'd probably be better off designing a whole new component (approach 2), taking inspiration from the existing code where relevant.

**Comment by Stannislav** [comment]
Hi @svlandeg that's for sure helpful.

Do you maybe also happen to know if there are any plans at explosion for changing to something different, or is this going to be the default NER model in the forseeable future? I'm asking because of the following points I mentioned in the summary above:

The current default model for NER is based on the model proposed in [Lample et al.](https://arxiv.org/abs/1603.01360) in 2016.

Given that it's already 5 years old I'm curious why there isn't a more recent model. What is the main motivation?
- This is still state of the art?
- It's fast?
- It can be re-used for both NER and dependency parsing?
- It's easy to implement checks for invalid IOB sequences?
- Something else?
